annotations/upload_validation.py

```diff
@@ -0,0 +1,89 @@
+'''
+upload_validation.py
+author: aditi.gopalan
+Description: This script checks if manifests have been uploaded to Synapse 
+'''
+
+import synapseclient
+import pandas as pd
+import csv
+import argparse
+from datetime import datetime
+
+def get_folder_type_argument():
+    parser = argparse.ArgumentParser(description="Process Synapse folders and create a manifest.")
+    parser.add_argument("csv_file_path", help="Path to the input CSV file.")
+    parser.add_argument("folder_type", choices=["Publication", "Dataset", "Tool"],
+                        help="Specify folder type: 'Publication', 'Dataset', or 'Tool'")
+    args = parser.parse_args()
+    return args.csv_file_path, args.folder_type
+
+def get_folder_id_column(folder_type):
+    if folder_type == "Publication":
+        return "folderIdPublication"
+    elif folder_type == "Dataset":
+        return "folderIdDatasets"
+    elif folder_type == "Tool":
+        return "folderIdTools"
+
+def get_project_name(syn, folder_id):
+    try:
+        folder = syn.get(folder_id)
+        project = syn.get(folder.parentId)
+        return project.name
+    except Exception as e:
+        print(f"Error getting information for {folder_id}: {e}")
+        return 'ERROR'
+
+def main():
+    csv_file_path, folder_type = get_folder_type_argument()
+    
+    syn = synapseclient.Synapse()
+    syn.login()
+
+    df = pd.read_csv(csv_file_path)
+    results = []
+
+    for index, row in df.iterrows():
+        folder_synapse_id = str(row[get_folder_id_column(folder_type)])
+        
+        if pd.notna(folder_synapse_id) and folder_synapse_id.lower() != 'nan':
+            project_name = get_project_name(syn, folder_synapse_id)
+            if project_name != 'ERROR':
+                try:
+                    folder_children = syn.getChildren(folder_synapse_id)
+                    most_recent_file = max(folder_children, key=lambda x: x['modifiedOn'])
+                    modification_time = most_recent_file.get('modifiedOn', 'N/A')
+                    results.append({'SynapseID': folder_synapse_id, 'ProjectName': project_name, 'ModifiedOn': modification_time})
+                except Exception as e:
+                    print(f"Error getting information for {folder_synapse_id}: {e}")
+                    results.append({'SynapseID': folder_synapse_id, 'ProjectName': 'ERROR', 'ModifiedOn': 'ERROR'})
```

Comment:
You can also alternatively add this dict to `results` as a Data Frame, e.g.

```python
results = []
for index, row in df.iterrows():
   ...
   results.append(pd.DataFrame({
      'SynapseID': [folder_synapse_id],
      'ProjectName': [project_name],
      'ModifiedOn': [modification_time]
   }))
```

And then combine all the DataFrames in `results` into a single DataFrame:

```python
new_results = pd.concat(results)
```

Now, later in the script, you can use `new_results.to_csv(...)`, that is, if you prefer using that over `csv.DictWriter()`.

Reply:
@vpchung thank you! I'm curious if this approach faster or just better coding practice?

Reply:
Ooo, that's a good question!  Looking at some benchmarks, `csv.DictWriter()` _is_ faster, so you're right - it would be totally ideal for large CSVs!  For smaller CSVs though, I don't think it will really matter and it's all based on preference.  You can probably tell that pandas is mine 😆 

Overall, your approach is more robust, considering the chance of having to print large CSVs.  Nicely done!!

Reply:
Thank you! This is helpful